### PR TITLE
fix(github-recon): handle Anthropic timeout and improve HTML extraction resilience

### LINKED ISSUE #115: fix(github-recon): handle Anthropic timeout and improve HTML extraction resilience
## Description

During GitHub repository recon (`sketchScreen`), two issues are causing otherwise valid recon jobs to fail:

1. **Anthropic SDK timeout validation** when using a large `max_tokens` value.
2. **Brittle HTML extraction** when the model returns valid HTML without a doctype, as a component fragment, or inside an incomplete/generic code fence.

Both issues affect the reliability of the GitHub recon pipeline and can result in cards being marked as failed even though the model response is usable.

---

## Issue 1: Anthropic SDK non-streaming timeout

When `sketchScreen()` uses:

`MODEL_MAX_TOKENS = 32_000`

the Anthropic SDK calculates an expected generation duration that exceeds its default non-streaming limit.

Because the Anthropic client is currently initialized without an explicit timeout, the SDK throws the following error before the request is dispatched:

```text
Streaming is required for operations that may take longer than 10 minutes.
```

### Relevant files

* `server/agentModel.ts`
* `server/distill.ts`
* `server/githubRecon.ts`

### Expected behavior

Long-running non-streaming requests should be allowed to execute within a configurable timeout.

### Proposed fix

Initialize the Anthropic client with an explicit, configurable timeout (for example, 15 minutes), ideally controlled through an environment variable such as:

`ANTHROPIC_TIMEOUT_MS`

This should be applied consistently wherever the Anthropic client is initialized.

---

## Issue 2: `extractHtml` is too strict

`extractHtml()` currently assumes that the model response contains a fenced HTML document with a `<!DOCTYPE html>` declaration.

In practice, the model can return several valid formats, including:

* HTML inside an `html` or `xml` code fence
* HTML inside a generic code fence
* An incomplete/unclosed code fence when generation is truncated
* `<html>...</html>` without `<!DOCTYPE html>`
* A component fragment such as `<div>`, `<main>`, `<section>`, or `<style>`
* Explanatory text or other fenced content before the actual HTML

The current extraction logic can therefore select the wrong code block or reject usable HTML entirely.

### Error

```text
Error: the model returned no HTML document
```

### Relevant file

* `server/githubRecon.ts`

### Expected behavior

`extractHtml()` should be resilient to the different formats the model may produce while still rejecting responses that contain no meaningful HTML.

For component recon, HTML fragments should be accepted and wrapped in a minimal document when necessary.

### Proposed fix

Update `extractHtml()` to:

* Prefer explicitly marked HTML/XML blocks.
* Detect HTML in generic code fences.
* Handle unclosed code fences.
* Fall back to searching the complete response when necessary.
* Accept `<html>` documents without a doctype and add one.
* Accept common component fragments and wrap them in a minimal HTML document.
* Continue throwing an error when the response genuinely contains no HTML.
* Preserve the existing height extraction behavior.

The detailed implementation can be kept in the PR rather than duplicated here.

---

## Impact

These issues can cause GitHub recon cards to fail for reasons unrelated to the actual quality of the generated reconstruction.

Fixing them should improve:

* Reliability of long-running recon requests
* Success rate of `sketchScreen()`
* Handling of component-level reconstructions
* Resilience to truncated or inconsistently formatted model output
* Debuggability of genuine extraction failures

---

## Acceptance Criteria

* [ ] Long `sketchScreen()` requests with `32k` max tokens no longer fail due to the Anthropic SDK's default non-streaming timeout validation.
* [ ] Anthropic timeout is configurable through an environment variable.
* [ ] `extractHtml()` correctly handles HTML/XML and generic code fences.
* [ ] Unclosed code fences do not automatically cause extraction to fail.
* [ ] HTML without `<!DOCTYPE html>` is accepted.
* [ ] Component HTML fragments are accepted and normalized into a renderable document.
* [ ] Invalid responses that contain no HTML still produce a clear error.
* [ ] Existing `doop-height` handling remains unchanged.

## Environment

* OS: Windows
* Runtime: Bun

## Files affected

* `server/githubRecon.ts`
* `server/agentModel.ts`
* `server/distill.ts`

### PR DESCRIPTION
**Pull request by Pallavikumarimdb**
## Description

Addresses two reliability issues during GitHub repository recon (`sketchScreen`):

1. **Anthropic SDK Timeout**: Configures `ANTHROPIC_TIMEOUT_MS` (default: 15m) when instantiating the Anthropic client in `agentModel.ts` and `distill.ts` to prevent `@anthropic-ai/sdk` from throwing `Streaming is required for operations that may take longer than 10 minutes` on 32k `maxTokens` requests.
2. **HTML Extraction Resilience**: Updates `extractHtml()` in `githubRecon.ts` to handle generic/XML fences, unclosed fences, missing `<!DOCTYPE>`, and UI component fragments (wrapping them into a minimal renderable HTML document).

## Changes
- `server/agentModel.ts`: Pass `timeout: ANTHROPIC_TIMEOUT_MS` to `new Anthropic()`.
- `server/distill.ts`: Use `ANTHROPIC_TIMEOUT_MS` for consistency.
- `server/githubRecon.ts`: Robust HTML extraction & component fallback wrapping in `extractHtml()`.
- `tests/githubRecon.test.ts`: Added unit tests for edge cases.

## Testing
- Ran `bun test tests/githubRecon.test.ts` (all 14 tests pass).
- Verified typechecking (`tsc --noEmit`) and linting/formatting pass cleanly.

Closes #115

### CODE REVIEW

File: server/agentModel.ts [hunk 1]
```diff
@@ -54,6 +54,7 @@ export { ModelAuthError }
 /* ---------------------------------------------------------------- */
 
 const ANTHROPIC_MODEL = process.env.DOOP_AGENT_MODEL || 'claude-opus-5'
+const ANTHROPIC_TIMEOUT_MS = Number(process.env.ANTHROPIC_TIMEOUT_MS) || 15 * 60 * 1000
```
[the remark this answers is not in the document]
Comment (Pallavikumarimdb) [reply server/agentModel.ts]: [b44fe45](https://github.com/kgoedecke/doop/pull/123/commits/b44fe456177264ff663149297b3304e6da9b3b44)

File: server/githubRecon.ts [hunk 2]
```diff
@@ -205,11 +205,50 @@ export function extractHtml(blocks: { type: string; text?: string }[]): { html:
...
-  if (start === -1) throw new Error('the model returned no HTML document')
-  html = html.slice(start)
+
+  let candidate = text.trim()
+  // 1. Check for ```html ... ``` or ```xml ... ``` code fences
+  const htmlFence = text.match(/```(?:html|xml)?\s*([\s\S]*?)(?:```|$)/i)
+  if (htmlFence && (htmlFence[1].includes('<') || htmlFence[1].toLowerCase().includes('doctype'))) {
+    candidate = htmlFence[1].trim()
+  } else {
+    // 2. Check for any code fence containing HTML tags
```
[the remark this answers is not in the document]
Comment (Pallavikumarimdb) [reply server/githubRecon.ts]: [b44fe45](https://github.com/kgoedecke/doop/pull/123/commits/b44fe456177264ff663149297b3304e6da9b3b44)

**Comment by kgoedecke** [comment]
Thanks for the PR. Both halves of #115 have now landed or are in flight through separate PRs that were opened a day earlier: #119 (merged) handles `extractHtml` accepting doctype-less documents and fragments, and #130 handles the SDK's long-request refusal by streaming the turn rather than raising the client timeout. With #119 anchoring fragment detection to a line start, the multi-fence priority scheme here isn't needed, and Greptile's note about an unrelated leading fence still shadowing a later document applies to it. Closing as superseded — appreciate the effort.
